### PR TITLE
Generify search form

### PR DESCRIPTION
**Pull request by jamesgorrie**
I had two options here:
* Pull out a new version of the search form and develop it under a toggle
* Refactor the current search form

I settled on the latter so we could see the diffs of changes going on, although I question if that has been the right choice. If people would prefer, I can do the former. Thoughts?

If I could get a general review (happy to take nits if you want too), it will give me an indication as to how people might prefer me to move forward. 

---

Generifies the `SearchForm` component so that you pass it config rather than it being aware of the images/works props.

This was triggered as when @ishouty came to add the filters for contributors/subjects/genres, it was a major code change, although much of the code had been written, just not generified.

As we are coming to adding new filters to the images search, I hope this work will make that easier.

We still have the concept of work & images to some degree because of the way the search tabbing is built, this would be the next point of call to refactor out.

Much of the code is in _working_ order, there is still a lot of tidy up, but I am trying to touch as little as possible to make the code PR to read (see first comment in description).

### CODE REVIEW

File: catalogue/webapp/pages/works.tsx [hunk 1]
```diff
@@ -87,25 +91,140 @@ const Works: NextPage<Props> = ({
...
+      works?.aggregations?.locationType?.buckets.map((bucket) => ({
+        id: bucket.data.type,
+        value: bucket.data.type,
+        count: bucket.count,
+        label: bucket.data.label,
+        selected: worksRouteProps.itemsLocationsType.includes(bucket.data.type),
+      })) || [],
+  };
+
+  const filters = [
```
Comment (jamieparkinson) [inline catalogue/webapp/pages/works.tsx:202]: Perhaps these could live in a separate file that provides `getFilters(works)` or something like that? `works.tsx` is a bit of a monster

File: catalogue/webapp/pages/works.tsx [hunk 2]
```diff
@@ -299,16 +411,20 @@ export const getServerSideProps: GetServerSideProps<
...
   const _queryType = cookies(context)._queryType;
-  const worksApiProps = worksRouteToApiUrl(params, {
+
+  const worksApiProps = worksRouteToApiUrl(props, {
     _queryType,
     aggregations,
   });
-  const shouldGetWorks = !!(params.query && params.query !== '');
+
+  const shouldGetWorks = !!(props.query && props.query !== '');
```
Comment (jamieparkinson) [inline catalogue/webapp/pages/works.tsx:426]: This is equivalent to `!!props.query` as empty strings are falsy

File: common/utils/routes.ts [hunk 3]
```diff
@@ -90,3 +100,48 @@ export function toSource<T extends string>(
...
+    }
+
+    // any empty values, we don't add
+    if (val === null || val === undefined || val === '') {
+      return { ...acc };
+    }
+
+    // As all our services default `page: undefined` to `page: 1` we remove it
+    const pageKeys = ['page', 'canvas', 'manifest'];
+    // console.info(key, val, typeof val);
```
Comment (jamieparkinson) [inline common/utils/routes.ts:137]: oop

File: common/utils/utility-types.ts [hunk 4]
```diff
@@ -24,3 +24,13 @@ export type OptionalToUndefined<T> = {
 export type ElementFromComponent<C extends FunctionComponent> = ReactElement<
   ComponentProps<C>
 >;
+
+/** Allows you to set a string type that requires a certain prefix
```
Comment (jamieparkinson) [inline common/utils/utility-types.ts:28]: This is cool

File: common/views/components/SearchForm/SearchForm.tsx [hunk 5]
```diff
@@ -136,74 +124,38 @@ const SearchForm: FunctionComponent<Props> = ({
...
+    // @ts-ignore
+    const url = new URLSearchParams(formData);
 
-    // Location type
-    const itemsLocationsType = (
-      nodeListValueToArray(form['items.locations.type']) || []
-    )
-      .filter(checkbox => checkbox.checked)
-      .map(checkbox => checkbox.value);
+    const params: ParsedUrlQuery = Array.from(url.entries()).reduce(
```
Comment (jamieparkinson) [inline common/views/components/SearchForm/SearchForm.tsx:139]: Can this live in a utilities file + have some comments explaining what it does?

File: common/views/components/SearchTabs/SearchTabs.tsx [hunk 6]
```diff
@@ -80,33 +73,35 @@ const Tab = styled(Space).attrs({
 const TabPanel = styled(Space)`
   background: ${(props) => props.theme.color('cream')};
 `;
+
 type Props = {
-  worksRouteProps: WorksRouteProps;
-  imagesRouteProps: ImagesRouteProps;
-  workTypeAggregations: CatalogueAggregationBucket[];
+  query: string;
+  sort: string | undefined;
```
Comment (jamieparkinson) [inline common/views/components/SearchTabs/SearchTabs.tsx:79]: Is there a reason these can't be `sort?: string`?

File: common/views/components/SearchTabs/SearchTabs.tsx [hunk 7]
```diff
@@ -168,13 +163,25 @@ const SearchTabs: FunctionComponent<Props> = ({
...
+            sort={sort}
+            sortOrder={sortOrder}
+            linkResolver={(params) => {
+              const query = propsToQuery(params);
+              const link = {
+                pathname: '/works',
+                query,
+              };
+
+              // setSavedWorksSearchState(fromQuery(query));
```
Comment (jamieparkinson) [inline common/views/components/SearchTabs/SearchTabs.tsx:176]: ?

File: common/views/components/SearchTabs/SearchTabs.tsx [hunk 8]
```diff
@@ -239,13 +246,23 @@ const SearchTabs: FunctionComponent<Props> = ({
             more.
           </Space>
           <SearchForm
+            query={query}
+            sort={undefined}
+            sortOrder={undefined}
+            linkResolver={(params) => {
+              const query = propsToQuery(params);
```
Comment (jamieparkinson) [inline common/views/components/SearchTabs/SearchTabs.tsx:253]: should this use `imagesLink` or similar?

File: common/hooks/useSavedSearchState.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { useState, useEffect, Dispatch, SetStateAction } from 'react';
+import { WorksProps } from '../views/components/WorksLink/WorksLink';
+
+let savedState;
```
Comment (jamieparkinson) [inline common/hooks/useSavedSearchState.ts:4]: I see why/how this works but it does seem _weird_ and I imagine changes in SSR could break it unexpectedly. Not sure what else to suggest though, perhaps something could be done using context APIs?

Separately I'm not sure about "saved" - I guessed that it was going to use localstorage, maybe `usePreviousSearchState`? Again not sure on a better option.

Comment (jamieparkinson) [reply common/hooks/useSavedSearchState.ts:4]: Reasonably confident that context is the "right" way to do it without mutation and closure etc etc, but if you want to get this PR in first and make that change next (I am also happy to have a go at it) that makes sense

Comment (jamesgorrie) [reply common/hooks/useSavedSearchState.ts:4]: Na, we had it like that before. I'll change it back.

File: catalogue/webapp/pages/works.tsx [hunk 10]
```diff
@@ -87,25 +91,140 @@ const Works: NextPage<Props> = ({
...
   }, []);
 
+  const productionDatesFilter: DateRangeFilter = {
+    type: 'dateRange',
+    id: 'production.dates',
+    label: 'Dates',
+    from: {
+      key: 'productionDatesFrom',
+      id: 'production.dates.from',
+      value: productionDatesFrom ?? undefined,
```
[the remark this answers is not in the document]
Comment (jamesgorrie) [reply catalogue/webapp/pages/works.tsx:101]: There definitely is, I think that might be part of the post-working tidy.

File: catalogue/webapp/pages/works.tsx (hunk 1 above)
Comment (jamesgorrie) [reply catalogue/webapp/pages/works.tsx:202]: I personally don't have a problem with working in the same file (as in the filters will only ever change at this point and aren't shared anywhere), I often prefer it, but if you think it makes it more readable, I'll move the filters (types and all) into it's own thing.

File: catalogue/webapp/components/WorksSearchResults/WorksSearchResults.tsx [hunk 11]
```diff
@@ -37,10 +37,10 @@ const WorkSearchResults: FunctionComponent<Props> = ({
                 position: i,
                 resultWorkType: result.workType.label,
                 resultIdentifiers: result.identifiers.map(
-                  identifier => identifier.value
+                  (identifier) => identifier.value
```
Comment (davidpmccormick) [inline catalogue/webapp/components/WorksSearchResults/WorksSearchResults.tsx:40]: nit: Prettier funkiness (I think my editor would remove these parens)

File: catalogue/webapp/pages/works.tsx (hunk 1 above)
Comment (davidpmccormick) [reply catalogue/webapp/pages/works.tsx:207]: I think I'd probably come down on the side of having them in a separate file. But I wouldn't be too upset if they weren't.

File: common/views/components/SearchTabs/SearchTabs.tsx (hunk 7 above)
Comment (jamesgorrie) [reply common/views/components/SearchTabs/SearchTabs.tsx]: removed

File: common/views/components/SearchTabs/SearchTabs.tsx (hunk 6 above)
Comment (jamesgorrie) [reply common/views/components/SearchTabs/SearchTabs.tsx]: Nope, that's been changed.

File: catalogue/webapp/pages/works.tsx [hunk 12]
```diff
@@ -141,24 +260,13 @@ const Works: NextPage<Props> = ({
...
-                    }}
-                    workTypeAggregations={
-                      works && works.aggregations
-                        ? works.aggregations.workType.buckets
-                        : []
-                    }
+                    query={worksRouteProps.query}
+                    sort={worksRouteProps.sort}
+                    sortOrder={worksRouteProps.sortOrder}
+                    worksFilters={filters}
```
Comment (davidpmccormick) [inline catalogue/webapp/pages/works.tsx:266]: Definitely feels quite a bit more manageable like this 👍

File: common/views/components/ModalFilters/ModalFilters.tsx [hunk 13]
```diff
@@ -1,279 +0,0 @@
-import {
```
Comment (davidpmccormick) [inline common/views/components/ModalFilters/ModalFilters.tsx:1]: I can't work out what's happening to this file. Is it being deleted and GitHub's just being a weirdo?

File: common/views/components/SearchFiltersDesktop/SearchFiltersDesktop.tsx [hunk 14]
```diff
@@ -235,100 +246,82 @@ const SearchFiltersDesktop: FunctionComponent<SearchFiltersSharedProps> = ({
...
-          subjectsSelected={subjectsSelected}
-          genresSelected={genresSelected}
-          contributorsSelected={contributorsSelected}
-        />
-      )}
+      <ResetActiveFilters
+        query={query}
+        linkResolver={linkResolver}
+        resetFilters={{
+          as: { pathname: '???', query: {} },
```
Comment (davidpmccormick) [inline common/views/components/SearchFiltersDesktop/SearchFiltersDesktop.tsx:320]: Are these `???` deliberate?

File: common/views/components/SearchForm/SearchForm.tsx (hunk 5 above)
Comment (davidpmccormick) [reply common/views/components/SearchForm/SearchForm.tsx:139]: 🙌

File: common/views/components/SearchTabs/SearchTabs.tsx [hunk 15]
```diff
@@ -2,27 +2,20 @@ import BaseTabs, { TabType } from '../BaseTabs/BaseTabs';
 import { classNames, font } from '@weco/common/utils/classnames';
 import styled from 'styled-components';
 import Space from '../styled/Space';
-import { useContext, FunctionComponent, ReactElement } from 'react';
+import React, { useContext, FunctionComponent, ReactElement } from 'react';
```
Comment (davidpmccormick) [inline common/views/components/SearchTabs/SearchTabs.tsx:5]: Couldn't see where/how you were using `React` in this file

**Review by davidpmccormick — COMMENTED** [review]
I fully back this approach 💯

File: common/views/components/SearchTabs/SearchTabs.tsx (hunk 8 above)
Comment (jamesgorrie) [reply common/views/components/SearchTabs/SearchTabs.tsx:251]: For now I've just left it as URLParams.

File: common/views/components/SearchForm/SearchForm.tsx (hunk 5 above)
Comment (jamesgorrie) [reply common/views/components/SearchForm/SearchForm.tsx:140]: I've done this, I'll get on some tests after this PR.

File: common/utils/utility-types.ts (hunk 4 above)
Comment (jamesgorrie) [reply common/utils/utility-types.ts:28]: It's the reason for the typescript bump.

File: common/views/components/SearchTabs/SearchTabs.tsx (hunk 15 above)
Comment (jamesgorrie) [reply common/views/components/SearchTabs/SearchTabs.tsx:5]: removed

File: catalogue/webapp/pages/works.tsx (hunk 1 above)
Comment (jamesgorrie) [reply catalogue/webapp/pages/works.tsx]: These have been moved into the [`services/catalogue/filters`](https://github.com/wellcomecollection/wellcomecollection.org/pull/6101/files#diff-2c7db8d4fc4625288df1d7d3db38c386271720e8e06b8efca385dd47dabcee41) file

File: common/views/components/SearchTabs/SearchTabs.tsx (hunk 8 above)
Comment (jamesgorrie) [reply common/views/components/SearchTabs/SearchTabs.tsx]: Actually, changed tact
https://github.com/wellcomecollection/wellcomecollection.org/pull/6101/files#diff-57484c78334523af4b4550d1f0423cb8395ef49a3ad6a67d4eb36333abfa8788R178

https://github.com/wellcomecollection/wellcomecollection.org/pull/6101/files#diff-57484c78334523af4b4550d1f0423cb8395ef49a3ad6a67d4eb36333abfa8788R178

Not the tidiest, but works.

File: common/views/components/ModalFilters/ModalFilters.tsx (hunk 13 above)
Comment (jamesgorrie) [reply common/views/components/ModalFilters/ModalFilters.tsx:1]: We don't use it anywhere. It was part of a toggle that went stale.

File: common/views/components/SearchFiltersDesktop/SearchFiltersDesktop.tsx (hunk 14 above)
Comment (jamesgorrie) [reply common/views/components/SearchFiltersDesktop/SearchFiltersDesktop.tsx]: No, nice pickup, they've been removed.

File: common/hooks/useSavedSearchState.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { useState, useEffect, Dispatch, SetStateAction } from 'react';
```
Comment (jamieparkinson) [inline common/hooks/useSavedSearchState.ts:1]: You can delete this file now I think

**Review by jamieparkinson — APPROVED** [review]
:tada:

**Pull request merged**